### DISCUSSION TITLE
How do I get the parameters I need when I request a file from the clipboard?

### DISCUSSION THREAD

**Discussion by kimcoldwater** [FreeRDP/FreeRDP#8827] [Q&A] [answered]
hello,

In a file clipboard request, you want to obtain a `FILEDESCRIPTOW* `structure from `ServerFormatDataResponse` and use the value to create a `CLIPRDR_FILE_CONTENTS_REQUEST` structure to send a `ClientFileContentsRequest`.

I would like to know the formula for obtaining `nPositionLow` and `nPositionHigh`, `cbRequested` according to dwFlags of `CLIPRDR_FILE_CONTENTS_REQUEST`.

The dwFlags are usually 1 (FILECONTENTS_SIZE),
How can I calculate the file data from ServerFileContentsResponse?

```
	CLIPRDR_FILE_CONTENTS_REQUEST fileContentsRequest;

	fileContentsRequest.streamId = 
	fileContentsRequest.listIndex = 
	fileContentsRequest.dwFlags = 
	fileContentsRequest.nPositionLow = 
	fileContentsRequest.nPositionHigh = 
	fileContentsRequest.cbRequested = 
	fileContentsRequest.clipDataId = 
	fileContentsRequest.msgFlags = 
```
I want to fill this structure.

**Comment by akallabeth** [comment, marked as answer]
* File stream to use `streamId`
* File to request data from: `listIndex` (the index in the `FILEDESCRIPTORW` list you got) 
* offset in file to read: `nPositionLow` and `nPositionHigh` 
* size to read from file: `cbRequested`
* (optional) if clipboard is locked, set `clipDataId` to the ID of the lock

the rest is not to be touched on send, it will contain data if received. (success/failure/...)

check `cliprdr_write_file_contents_request` for a use of that.

**Reply by kimcoldwater** [reply]
If you copy a file during a connection, the server `FormatDataResponse` imports the data in `FileGroupDescriptorW` format. However, is it different from the information here at `https://learn.microsoft.com/en-us/windows/win32/api/shlobj_core/ns-shlobj_core-filedescriptorw `that it maps to `FILEDESCRIPTORW` rather than `FILEDESCRIPTORW`, but `FILEDESCRIPTORW` is defined in freerdp?

```
if ( strcmp( ClipboardGetFormatName(tfc->clipboard, formatId), "FileGroupDescriptorW\0") == 0){

		formatId = ClipboardGetFormatId(clipboard, "FileGroupDescriptorW");

		data = (void *)ClipboardGetData(clipboard, formatId, &size);

		FILEDESCRIPTORW* groupDsc = (FILEDESCRIPTORW*)data;

		UINT32 file_count = size / sizeof(FILEDESCRIPTORW);
}
```

Cannot map to `FILEGROUPDESCRIPTOW*` structure.

**Reply by akallabeth** [reply]
ok, sorry, you mix up something there. The procedure is:

1. The server (or client) announces a new format with type `FileGroupDescriptorW` that you need to get from the remote end. (With the `FormatList` message)
2. you need to get the format data with `FormatDataRequest` 
3. you get a `FormatDataResponse`, you need to parse that, see `cliprdr_parse_file_list` on how to extract data, to know which files are available.
5. only then you can start with requesting file size or contents with `FileContentsRequest` and `FileContentsResponse`

**Reply by kimcoldwater** [reply]
Thank you. Thanks to your advice, I got the file.

Is this the right method to send the ClientFileContentsRequest as many as the number of divisions when requesting a split request?

**Reply by akallabeth** [reply]
sorry, what are you trying to ask?
you send a request for anything you´d like to get, it is up to you to divide the sizes each request returns. (with one exception, if the file is `>4GB` then you need to split as one request has limits)

**Reply by kimcoldwater** [reply]
For example, when I want to receive a file of 1GB and want to request a split of 100MB, I wonder if it is the right way to send 10 requests of 100MB each using nPositionLow, nPositioHigh, and cbRequested in ClientFileContentsRequest.

**Reply by akallabeth** [reply]
@kimcoldwater you´ll need to try out where your sweetspot between overhead and latency is. (e.g. if you request `100MB` on your slow connection the transfer might not abort before the data is transferred even if you cancel the copy), but other than that, yes, that is the way.

**Reply by kimcoldwater** [reply]
OK, now the contents of the 'Server->Client' file are successfully imported and printed.
thank you.
